src/tag.rs

```diff
@@ -267,9 +349,29 @@ impl Tag {
             return Ok(None)
         }
         let byte = source.slice()[0];
-        let (tag, compressed) = (Tag(byte & 0xdf), byte & 0x20 != 0);
+        // clear constructed bit
+        let mut data = [byte & !Tag::CONSTRUCTED_MASK, 0, 0, 0];
+        if (data[0] & Tag::SINGLEBYTE_DATA_MASK) == Tag::SINGLEBYTE_DATA_MASK {
+            let mut i = 1;
+            loop {
+                if source.request(i+1)? == 0 {
+                    // TODO: Should we return an error instead?
+                    return Ok(None)
+                }
+                data[i] = source.slice()[i];
+                if data[i] & Tag::LAST_OCTET_MASK == 0 {
+                    break
+                }
+                // We don’t support tags larger than 4 bytes.
+                if i == 3 {
+                    xerr!(return Err(decode::Error::Unimplemented.into()))
+                }
+                i += 1;
+            }
+        }
+        let (tag, compressed) = (Tag(data), byte & Tag::CONSTRUCTED_MASK != 0);
         if tag == self {
```

Comment:
This if can probably be merged into the loop now?

Reply:
I could not find a clean way since the first byte case is outside the loop, and I'm not sure how to bring it in.